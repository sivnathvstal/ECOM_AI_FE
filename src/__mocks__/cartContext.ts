import { Cart, CartProduct } from '../types';

export const mockCartItems: CartProduct[] = [
  {
    productId: 1,
    quantity: 2,
    product: {
      id: 1,
      title: 'Test Product 1',
      price: 10.99,
      description: 'Test description 1',
      category: 'electronics',
      image: 'test1.jpg',
      rating: { rate: 4.5, count: 100 },
    },
  },
  {
    productId: 2,
    quantity: 1,
    product: {
      id: 2,
      title: 'Test Product 2',
      price: 20.99,
      description: 'Test description 2',
      category: 'clothing',
      image: 'test2.jpg',
      rating: { rate: 4.0, count: 80 },
    },
  },
];

export const mockCart: Cart = {
  id: 1,
  userId: 1,
  date: '2023-01-01',
  products: [
    { productId: 1, quantity: 2 },
    { productId: 2, quantity: 1 },
  ],
};

export const mockCartContext = {
  cart: mockCart,
  cartItems: mockCartItems,
  isLoading: false,
  error: null,
  addItem: jest.fn(),
  updateItem: jest.fn(),
  removeItem: jest.fn(),
  totalItems: 3, // 2 + 1
  totalPrice: 42.97, // 10.99*2 + 20.99*1
};

export const mockEmptyCartContext = {
  cart: {
    id: 0,
    userId: 1,
    date: '2023-01-01',
    products: [],
  },
  cartItems: [],
  isLoading: false,
  error: null,
  addItem: jest.fn(),
  updateItem: jest.fn(),
  removeItem: jest.fn(),
  totalItems: 0,
  totalPrice: 0,
};

export const mockLoadingCartContext = {
  ...mockCartContext,
  isLoading: true,
};

export const mockErrorCartContext = {
  ...mockCartContext,
  error: 'Failed to fetch cart',
};

const defaultExport = {
  mockCartItems,
  mockCart,
  mockCartContext,
  mockEmptyCartContext,
  mockLoadingCartContext,
  mockErrorCartContext,
};
export default defaultExport; 