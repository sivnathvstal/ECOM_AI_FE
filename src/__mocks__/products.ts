import { Product, PaginationParams, ProductsResponse } from '../types';

export const mockProducts: Product[] = [
  {
    id: 1,
    title: 'Test Product 1',
    price: 10.99,
    description: 'Test description 1',
    category: 'electronics',
    image: 'test1.jpg',
    rating: { rate: 4.5, count: 100 },
  },
  {
    id: 2,
    title: 'Test Product 2',
    price: 20.99,
    description: 'Test description 2',
    category: 'clothing',
    image: 'test2.jpg',
    rating: { rate: 4.0, count: 80 },
  },
  {
    id: 3,
    title: 'Test Product 3',
    price: 15.99,
    description: 'Test description 3',
    category: 'electronics',
    image: 'test3.jpg',
    rating: { rate: 3.5, count: 60 },
  },
  {
    id: 4,
    title: 'Test Product 4',
    price: 25.99,
    description: 'Test description 4',
    category: 'jewelry',
    image: 'test4.jpg',
    rating: { rate: 4.8, count: 120 },
  },
];

export const mockPagination: PaginationParams = {
  page: 1,
  limit: 8,
  total: 4,
};

export const mockProductsResponse: ProductsResponse = {
  products: mockProducts,
  pagination: mockPagination,
};

export const mockCategories: string[] = [
  'electronics',
  'clothing',
  'jewelry',
]; 