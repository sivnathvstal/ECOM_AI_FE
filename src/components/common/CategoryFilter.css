.category-filter {
  margin: 8px;
  min-width: 200px;
  margin-bottom: 16px;
  width: 250px;
} 