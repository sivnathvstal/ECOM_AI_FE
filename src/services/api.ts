import axios from 'axios';

const API_URL = process.env.API_URL || 'https://fakestoreapi.com';

const api = axios.create({
  baseURL: API_URL,
  headers: {
    'Content-Type': 'application/json',
  },
});

export default api; 