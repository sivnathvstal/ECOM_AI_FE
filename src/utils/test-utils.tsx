import React, { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { ThemeProvider, createTheme } from '@material-ui/core';
import { QueryClient, QueryClientProvider } from 'react-query';
import { CartProvider } from '../context/CartContext';

// Create a theme for testing
const theme = createTheme({
  palette: {
    primary: {
      main: '#3f51b5',
    },
    secondary: {
      main: '#f50057',
    },
  },
});

// Create a client for React Query
const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
    },
  },
});

// All providers wrapper
const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <QueryClientProvider client={queryClient}>
      <ThemeProvider theme={theme}>
        <CartProvider>
          <BrowserRouter>{children}</BrowserRouter>
        </CartProvider>
      </ThemeProvider>
    </QueryClientProvider>
  );
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, ...options });

// Re-export everything
export * from '@testing-library/react';

// Override render method
export { customRender as render }; 