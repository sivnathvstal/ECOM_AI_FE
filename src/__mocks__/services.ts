import { mockProducts, mockProductsResponse, mockCategories } from './products';
import { mockCart, mockCartItems } from './cartContext';

// Product service mocks
export const mockProductService = {
  getProducts: jest.fn().mockResolvedValue(mockProductsResponse),
  getProductById: jest.fn().mockImplementation((id: number) => {
    const product = mockProducts.find(p => p.id === id);
    if (product) {
      return Promise.resolve(product);
    }
    return Promise.reject(new Error('Product not found'));
  }),
  getCategories: jest.fn().mockResolvedValue(mockCategories),
  searchProducts: jest.fn().mockResolvedValue(mockProductsResponse),
};

// Cart service mocks
export const mockCartService = {
  getCart: jest.fn().mockResolvedValue(mockCart),
  addToCart: jest.fn().mockResolvedValue(mockCart),
  updateCartItem: jest.fn().mockResolvedValue(mockCart),
  removeFromCart: jest.fn().mockResolvedValue(mockCart),
  getCartWithProducts: jest.fn().mockResolvedValue(mockCartItems),
}; 