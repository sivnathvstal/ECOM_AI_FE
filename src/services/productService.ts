import api from './api';
import { Product, ProductsResponse } from '../types';

export const getProducts = async (
  page = 1,
  limit = 8,
  category?: string
): Promise<ProductsResponse> => {
  try {
    // First get all products to calculate total
    const allProductsResponse = await api.get<Product[]>('/products');
    const allProducts = allProductsResponse.data;
    
    // Filter by category if provided
    const filteredProducts = category 
      ? allProducts.filter((product) => product.category === category) 
      : allProducts;
    
    const total = filteredProducts.length;
    
    // Calculate pagination
    const startIndex = (page - 1) * limit;
    const endIndex = startIndex + limit;
    const paginatedProducts = filteredProducts.slice(startIndex, endIndex);
    
    return {
      products: paginatedProducts,
      pagination: {
        page,
        limit,
        total
      }
    };
  } catch (error) {
    console.error('Error fetching products:', error);
    throw error;
  }
};

export const getProductById = async (id: number): Promise<Product> => {
  try {
    const response = await api.get<Product>(`/products/${id}`);
    return response.data;
  } catch (error) {
    console.error(`Error fetching product with id ${id}:`, error);
    throw error;
  }
};

export const getCategories = async (): Promise<string[]> => {
  try {
    const response = await api.get<string[]>('/products/categories');
    return response.data;
  } catch (error) {
    console.error('Error fetching categories:', error);
    throw error;
  }
};

export const searchProducts = async (
  query: string,
  page = 1,
  limit = 8
): Promise<ProductsResponse> => {
  try {
    const allProductsResponse = await api.get<Product[]>('/products');
    const allProducts = allProductsResponse.data;
    
    // Filter products by search query
    const filteredProducts = allProducts.filter((product) => 
      product.title.toLowerCase().includes(query.toLowerCase()) ||
      product.description.toLowerCase().includes(query.toLowerCase())
    );
    
    const total = filteredProducts.length;
    
    // Calculate pagination
    const startIndex = (page - 1) * limit;
    const endIndex = startIndex + limit;
    const paginatedProducts = filteredProducts.slice(startIndex, endIndex);
    
    return {
      products: paginatedProducts,
      pagination: {
        page,
        limit,
        total
      }
    };
  } catch (error) {
    console.error('Error searching products:', error);
    throw error;
  }
}; 