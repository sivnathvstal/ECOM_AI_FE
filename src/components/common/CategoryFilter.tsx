import React, { useEffect, useState } from 'react';
import {
  FormControl,
  InputLabel,
  Select,
  MenuItem
} from '@material-ui/core';
import { getCategories } from '../../services/productService';
import './CategoryFilter.css';

interface CategoryFilterProps {
  selectedCategory: string;
  onCategoryChange: (category: string) => void;
}

const CategoryFilter: React.FC<CategoryFilterProps> = ({ selectedCategory, onCategoryChange }) => {
  const [categories, setCategories] = useState<string[]>([]);
  const [loading, setLoading] = useState<boolean>(true);

  useEffect(() => {
    const fetchCategories = async () => {
      try {
        setLoading(true);
        const data = await getCategories();
        setCategories(data);
      } catch (error) {
        console.error('Error fetching categories:', error);
      } finally {
        setLoading(false);
      }
    };

    fetchCategories();
  }, []);

  const handleChange = (event: React.ChangeEvent<{ value: unknown }>) => {
    onCategoryChange(event.target.value as string);
  };

  return (
    <FormControl className="category-filter" data-testid="category-filter">
      <InputLabel id="category-select-label">Category</InputLabel>
      <Select
        labelId="category-select-label"
        id="category-select"
        value={selectedCategory}
        onChange={handleChange}
        disabled={loading}
        data-testid="category-select"
      >
        <MenuItem value="" data-testid="category-all">
          All Categories
        </MenuItem>
        {categories.map((category) => (
          <MenuItem 
            key={category} 
            value={category}
            data-testid={`category-${category}`}
          >
            {category.charAt(0).toUpperCase() + category.slice(1)}
          </MenuItem>
        ))}
      </Select>
    </FormControl>
  );
};

export default CategoryFilter; 