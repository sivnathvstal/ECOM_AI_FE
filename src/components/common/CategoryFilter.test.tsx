import React from 'react';
import { render, screen, fireEvent, waitFor, act, within } from '../../utils/test-utils';
import CategoryFilter from './CategoryFilter';
import * as productService from '../../services/productService';
import { mockCategories } from '../../__mocks__/products';

// Mock the productService
jest.mock('../../services/productService', () => ({
  getCategories: jest.fn(),
}));

describe('CategoryFilter', () => {
  const mockOnCategoryChange = jest.fn();
  
  beforeEach(() => {
    jest.clearAllMocks();
    (productService.getCategories as jest.Mock).mockResolvedValue(mockCategories);
  });

  it('renders correctly', async () => {
    let container;
    await act(async () => {
      const rendered = render(
        <CategoryFilter 
          selectedCategory="" 
          onCategoryChange={mockOnCategoryChange} 
        />
      );
      container = rendered.container;
    });
    await waitFor(() => expect(productService.getCategories).toHaveBeenCalled());
    expect(container).toMatchSnapshot();
  });

  it('fetches categories on mount', async () => {
    await act(async () => {
      render(
        <CategoryFilter 
          selectedCategory="" 
          onCategoryChange={mockOnCategoryChange} 
        />
      );
    });
    await waitFor(() => expect(productService.getCategories).toHaveBeenCalled());
    
    const select = screen.getByTestId('category-select');
    // Open the dropdown
    fireEvent.mouseDown(select.querySelector('[role="button"]') || select);
    // Wait for the listbox to appear
    const listbox = await screen.findByRole('listbox');
    // Wait for categories to be rendered
    await waitFor(() => {
      expect(within(listbox).getByTestId('category-all')).toBeInTheDocument();
      mockCategories.forEach(category => {
        expect(within(listbox).getByTestId(`category-${category}`)).toBeInTheDocument();
      });
    });
  });

  it('displays the selected category', async () => {
    const selectedCategory = 'electronics';
    await act(async () => {
      render(
        <CategoryFilter 
          selectedCategory={selectedCategory} 
          onCategoryChange={mockOnCategoryChange} 
        />
      );
    });
    
    await waitFor(() => expect(productService.getCategories).toHaveBeenCalled());
    const select = screen.getByTestId('category-select');
    fireEvent.mouseDown(select.querySelector('[role="button"]') || select);
    const listbox = await screen.findByRole('listbox');
    const selectedOption = within(listbox).getByText('Electronics');
    expect(selectedOption).toHaveAttribute('aria-selected', 'true');
  });

  it('calls onCategoryChange when a category is selected', async () => {
    await act(async () => {
      render(
        <CategoryFilter 
          selectedCategory="" 
          onCategoryChange={mockOnCategoryChange} 
        />
      );
    });
    
    await waitFor(() => expect(productService.getCategories).toHaveBeenCalled());
    
    const select = screen.getByTestId('category-select');
    // Open the dropdown
    fireEvent.mouseDown(select.querySelector('[role="button"]') || select);
    const listbox = await screen.findByRole('listbox');
    // Click on a category option
    const electronicsOption = within(listbox).getByTestId('category-electronics');
    fireEvent.click(electronicsOption);
    expect(mockOnCategoryChange).toHaveBeenCalledWith('electronics');
  });

  it('handles API error gracefully', async () => {
    const error = new Error('Failed to fetch categories');
    (productService.getCategories as jest.Mock).mockRejectedValue(error);
    
    const consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});
    
    await act(async () => {
      render(
        <CategoryFilter 
          selectedCategory="" 
          onCategoryChange={mockOnCategoryChange} 
        />
      );
    });
    
    await waitFor(() => expect(productService.getCategories).toHaveBeenCalled());
    // Open the dropdown
    const select = screen.getByTestId('category-select');
    fireEvent.mouseDown(select.querySelector('[role="button"]') || select);
    const listbox = await screen.findByRole('listbox');
    // Should still render the "All Categories" option
    expect(within(listbox).getByTestId('category-all')).toBeInTheDocument();
    // Should log the error
    expect(consoleErrorSpy).toHaveBeenCalledWith('Error fetching categories:', error);
    consoleErrorSpy.mockRestore();
  });
}); 