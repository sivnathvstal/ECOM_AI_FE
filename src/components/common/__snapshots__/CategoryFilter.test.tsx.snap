// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryFilter renders correctly 1`] = `
<div>
  <div
    class="MuiFormControl-root category-filter"
    data-testid="category-filter"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
      data-shrink="false"
      id="category-select-label"
    >
      Category
    </label>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      data-testid="category-select"
    >
      <div
        aria-haspopup="listbox"
        aria-labelledby="category-select-label category-select"
        class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
        id="category-select"
        role="button"
        tabindex="0"
      >
        <span>
          â€‹
        </span>
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput"
        tabindex="-1"
        value=""
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSelect-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
</div>
`;
